File: nagasenacs/src/Nagasena/Sax/TransmogrifierException.cs
using System;

namespace Nagasena.Sax {

  using Locator = Org.System.Xml.Sax.ILocator;
  using LocatorImpl = Org.System.Xml.Sax.Helpers.LocatorImpl;

  using MessageResolver = Nagasena.Util.MessageResolver;

  /// <summary>
  /// Exception handler for the Transmogrifier.
  /// </summary>
  public sealed class TransmogrifierException : Exception {

    private const long serialVersionUID = -4536662596727577640L;
    /// <summary>
    /// Unexpected Element.
    /// </summary>
    public const int UNEXPECTED_ELEM = 1;
    /// <summary>
    /// Unexpected Attribute.
    /// </summary>
    public const int UNEXPECTED_ATTR = 2;
    /// <summary>
    /// Unexpected Character Sequence.
    /// </summary>
    public const int UNEXPECTED_CHARS = 3;
    /// <summary>
    /// Unexpected Binary value.
    /// </summary>
    public const int UNEXPECTED_BINARY_VALUE = 4;
    /// <summary>
    /// Unhandled SAX parser feature.
    /// </summary>
    public const int UNHANDLED_SAXPARSER_FEATURE = 5;
    /// <summary>
    /// SAX error reported by XML parser.
    /// </summary>
    public const int SAX_ERROR = 6;
    /// <summary>
    /// Unexpected End of Element event.
    /// </summary>
    public const int UNEXPECTED_END_ELEM = 7;
    /// <summary>
    /// Unexpected End of Document event.
    /// </summary>
    public const int UNEXPECTED_ED = 8;
    /// <summary>
    /// Unexpected Start of Document event.
    /// </summary>
    public const int UNEXPECTED_SD = 9;
    /// <summary>
    /// Prefix is not bound.
    /// </summary>
    public const int PREFIX_NOT_BOUND = 10;
    /// <summary>
    /// Prefix is bound to another namespace.
    /// </summary>
    public const int PREFIX_BOUND_TO_ANOTHER_NAMESPACE = 11;
    /// <summary>
    /// Errors reported by Scriber.
    /// </summary>
    public const int SCRIBER_ERROR = 12;

    private static readonly MessageResolver m_msgs = new TransmogrifierExceptionMessages();

    private readonly int m_code;
    private readonly string m_message;

    private Exception m_exception = null;
    private Locator m_locator = null;

    /// <summary>
    /// Constructs a new TransmogrifierException. </summary>
    /// <param name="code"> int value  that represents the type of the exception </param>
    /// <param name="texts"> one or more strings that describe the exception </param>
    internal TransmogrifierException(int code, string[] texts) {
      m_code = code;
      m_message = m_msgs.getMessage(code, texts);
      m_locator = null;
    }

    /// <summary>
    /// Constructs a new TransmogrifierException. </summary>
    /// <param name="code"> int value  that represents the type of the exception </param>
    /// <param name="texts"> one or more strings that describe the exception </param>
    /// <param name="locator">  Locator for where the error occurred </param>
    internal TransmogrifierException(int code, string[] texts, LocatorImpl locator) {
      m_code = code;
      m_message = m_msgs.getMessage(code, texts);
      m_locator = locator;
    }

    /// <summary>
    /// Returns a code that represents the type of the exception. </summary>
    /// <returns> error code </returns>
    public int Code {
      get {
        return m_code;
      }
    }

    /// <summary>
    /// Returns a message that describes the exception. </summary>
    /// <returns> error message </returns>
    public override string Message {
      get {
        return m_message;
      }
    }

    /// <summary>
    /// Returns an Exception object. </summary>
    /// <returns> the error as an Exception instance </returns>
    public Exception Exception {
      get {
        return m_exception;
      }
      set {
        m_exception = value;
      }
    }


    /// <summary>
    /// Returns the locator that is associated with this compilation error. </summary>
    /// <returns> a Locator if available, otherwise null </returns>
    public Locator Locator {
      get {
        return m_locator;
      }
    }

  }

}

File: nagasena/src/org/openexi/sax/TransmogrifierException.java
package org.openexi.sax;

import org.xml.sax.Locator;
import org.xml.sax.helpers.LocatorImpl;

import org.openexi.util.MessageResolver;

/**
 * Exception handler for the Transmogrifier.
 */
public final class TransmogrifierException extends Exception {

  private static final long serialVersionUID = -4536662596727577640L;
  /**
   * Unexpected Element.
   */
  public static final int UNEXPECTED_ELEM    = 1;
  /**
   * Unexpected Attribute.
   */
  public static final int UNEXPECTED_ATTR    = 2;
  /**
   * Unexpected Character Sequence.
   */
  public static final int UNEXPECTED_CHARS   = 3;
  /**
   * Unexpected Binary value.
   */
  public static final int UNEXPECTED_BINARY_VALUE   = 4;
  /**
   * Unhandled SAX parser feature.
   */
  public static final int UNHANDLED_SAXPARSER_FEATURE = 5;
  /**
   * SAX error reported by XML parser.
   */
  public static final int SAX_ERROR = 6;
  /**
   * Unexpected End of Element event.
   */
  public static final int UNEXPECTED_END_ELEM = 7;
  /**
   * Unexpected End of Document event.
   */
  public static final int UNEXPECTED_ED = 8;
  /**
   * Unexpected Start of Document event.
   */
  public static final int UNEXPECTED_SD = 9;
  /**
   * Prefix is not bound.
   */
  public static final int PREFIX_NOT_BOUND = 10;
  /**
   * Prefix is bound to another namespace.
   */
  public static final int PREFIX_BOUND_TO_ANOTHER_NAMESPACE = 11;
  /**
   * Errors reported by Scriber.
   */
  public static final int SCRIBER_ERROR = 12;
  /**
   * Whitespaces could not be preserved.
   * This occurs when xml:space was specified to be preserved in strict mode.
   */
  public static final int CANNOT_PRESERVE_WHITESPACES = 13;
  /**
   * Errors reported by EXIOptionsEncoder.
   */
  public static final int EXI_OPTIONS_ENCODER_EXCEPTION = 14;

  private static final MessageResolver m_msgs =
    new MessageResolver(TransmogrifierException.class);

  private final int m_code;
  private final String m_message;
  
  private Exception m_exception = null;
  private Locator m_locator = null;
  
  /**
   * Constructs a new TransmogrifierException.  
   * @param code int value  that represents the type of the exception
   * @param texts one or more strings that describe the exception
   */
  TransmogrifierException(int code, String[] texts) {
    m_code    = code;
    m_message = m_msgs.getMessage(code, texts);
    m_locator = null;
  }
  
  /**
   * Constructs a new TransmogrifierException.  
   * @param code int value  that represents the type of the exception
   * @param texts one or more strings that describe the exception
   * @param locator  Locator for where the error occurred
   */
  TransmogrifierException(int code, String[] texts, LocatorImpl locator) {
    m_code    = code;
    m_message = m_msgs.getMessage(code, texts);
    m_locator = locator; 
  }
  
  /**
   * Returns a code that represents the type of the exception.
   * @return error code
   */
  public int getCode() {
    return m_code;
  }
  
  /**
   * Returns a message that describes the exception.
   * @return error message
   */
  @Override
  public String getMessage() {
    return m_message;
  }

  /**
   * Returns an Exception object.
   * @return the error as an Exception instance
   */
  public Exception getException() {
    return m_exception;
  }

  void setException(Exception exc) {
    m_exception = exc;
  }
  
  /**
   * Returns the locator that is associated with this compilation error.
   * @return a Locator if available, otherwise null
   */
  public Locator getLocator() {
    return m_locator;
  }

}
